### DISCUSSION TITLE
Running marimo on modal

### DISCUSSION THREAD

**Discussion by hgbrian** [marimo-team/marimo#2144] [Q&A]
Hi, I am trying to get marimo running on modal, following https://docs.marimo.io/guides/deploying/programmatically.html and https://docs.marimo.io/guides/deploying/authentication.html

The following code works, up to the point of giving a 401 for /api/kernel/instantiate

```python
import modal
from modal import asgi_app
import marimo

server = marimo.create_asgi_app().with_app(path="", root="/marimo/test.py")

app = modal.App()

@app.function(
    image=modal.Image.debian_slim().pip_install("marimo==0.8.3", "fastapi"),
    gpu=False,
    mounts=[modal.Mount.from_local_dir("./marimo", remote_path="/marimo")],
)
@asgi_app()
def marimo_asgi():
    return server.build()


if __name__ == "__main__":
    modal.serve(app)
```

I have tried adding middleware and I can see the requests come through.
I can reject requests (return 401) but I can't figure out how to prevent a 401 for /api/kernel/instantiate

```
    GET / -> 200 OK  (duration: 117.8 ms, execution: 51.9 ms)
    ....
    GET /android-chrome-192x192.png -> 200 OK  (duration: 1.27 s, execution: 0.0 ms)
   POST /api/kernel/instantiate -> 401 Unauthorized  (duration: 1.17 s, execution: 13.1 ms)
CONNECT /ws -> 101 Switching Protocols  (duration: 1.52 s, execution: 0.0 ms)
```

I think I am missing something fundamental about what's going on here, since I don't see where a token could be validated, or where the 401 is actually happening. I do see `marimo-server-token` in the header for /api/kernel/instantiate but no `authorization` or similar.

If anyone has any ideas, I would appreciate it! Running marimo on modal would be fantastic.

Example middleware code (note this is one of several attempts. I turn on and off the auth_header check, edit the headers, etc. and no matter how permissive, nothing changes):

```python
def auth_middleware(app):
    async def middleware(scope, receive, send):
        if scope['type'] == 'http':
            headers = dict(scope['headers'])
            auth_header = headers.get(b'authorization', b'').decode()
            expected_auth = f"Bearer {TOKEN}"

            if auth_header != expected_auth:
                print(f"Authentication failed. Expected: {expected_auth}, Got: {auth_header}")
                await send({
                    'type': 'http.response.start',
                    'status': 401,
                    'headers': [(b'content-type', b'text/plain')]
                })
                await send({
                    'type': 'http.response.body',
                    'body': b'Unauthorized\n'
                })
                return
        await app(scope, receive, send)

    return middleware
```

**Comment by akshayka** [comment]
for others interested, we followed up on discord: https://discord.gg/JE7nhX6mD8
